content/blog/2020-10-28-type-based-alias-analysis.md

```diff
@@ -0,0 +1,138 @@
++++
+title = "Type-Based Alias Analysis"
+[extra]
+bio = """
+  Kenneth Li is a first-semester M.Eng. student who is interested in functional
+  programming. He sings bass in the 
+  [Cornell Hangovers](http://www.hangovers.com/).
+"""
+[[extra.authors]]
+name = "Kenneth Li"
++++
+
+Alias analysis allows compilers to determine which pointers may (or must) 
+refer to the same memory location. This is most useful for the purpose of 
+instruction reordering; if a compiler knows that two memory instructions refer 
+to different memory locations, it can switch the order of the instructions. In 
+_Type-Based Alias Analysis_, Diwan, McKinley, and Moss propose and evaluate 
+alias analyses for Modula-3 (a type-safe language with in heritance) that use 
+type information instead of assignment tracking. Each analysis successively 
+refines the previous: _TypeDecl_ uses only type/subtype compatibility, 
+_FieldTypeDecl_ additionally considers field names and other properties, and 
+_SMFieldTypeRefs_ augments the former with a single pass through the 
+instructions to pick up assignment information. The analyses all do a may-alias 
+analysis; all alias pairs calculated are possible aliases, but all other pairs 
+are confirmed to not be aliases.
+
+## Algorithms
+The _TypeDecl_ analysis relies on a simple predicate: two referenced memory 
+access paths `p` and `q` may be aliases if and only if there exists a type that 
+subtypes both the type of `p` and the type of `q`. If such a type did not 
+exist, for example, then if `p` and `q`were aliases, the object they refer to 
+would have two different, incompatible types. _FieldTypeDecl_ takes this 
+further by adding checks for field names, array accesses, and pointer 
+dereferences. For example, paths `p.f` and `q.g` may be aliases only if the 
+field names `f` and `g` are exactly the same and `p` and `q` themselves may be 
+aliases. Additionally, a pointer dereference `q^` may alias a field `p.f` or an 

```

Comment:
As above, it's probably a good idea to define that `q^` means in the assumed language (because most readers will not have programmed in Modula-3, but each concept has a direct analog in mainstream 2020 languages).

Reply:
clarified with analogy to C's `*q` syntax